File: src/mapping/make_mapping.py
#!/usr/bin/python
# -*- coding: utf-8 -*-
# -*- vi: set ts=4 sts=4 sw=4 : -*-

import codecs
import os
import re
import subprocess
import sys

class category_component :

	category = []
	component = ""

	def __init__(self, category, component) :
		self.category = category
		self.component = component

	def category_to_string(self) :
		return " ".join(category)

	def equals_categories(self, category) :
		rc = category[0] == category[0] \
			and category[1] == category[1] \
			and category[2] == category[2] \
			and category[3] == category[3]
		return rc

	def get_component(self) :
		return self.component

	def get_category(self) :
		return self.category


class make_mapping :

	STATIC_MAPPING = {
		"..." : "...",
	}

	SKIP_COMPONENTS = [
		"...",
	]

	PATTERN1 = re.compile("^([\w\/]*)\/src\/(.*)$")
	PATTERN2 = re.compile("^([\S ]*)\t([\S ]*)\t([\S ]*)\t([\S ]*)\t([\w\.-]*)\t*$")
	PATTERN3 = re.compile("^\S*\.(sh|adf|sts|ear|jar)$")

	def debug(self, msg) :
#		print msg
		return

	def warning(self, msg) :
		print >> sys.stderr, "Warning: %s" % msg
		return

	def skip_component(self, component) :
		result3 = self.PATTERN3.match(component)
		if result3 :
			self.debug("skip_component: excluded component=[%s]" % component)
			return True
		for c in self.SKIP_COMPONENTS :
			if c == component :
				self.debug("skip_component: excluded component=[%s]" % component)
				return True
		return False

	def check_duplicate_unit(selt, list, unit, component, category) :
		if list.get(unit) :
			if component == list[unit].get_component() :
				if not list[unit].equals_categories(category) :
					p1 = " ".join(category)
					p2 = list[unit].categories_to_string()
					self.warning("duplicate components. (component=[%s] category=[%s] => [%s)" % (component, p1, p2))
				return False
			if not list[unit].equals_categories(category) :
				p1 = list[unit].get_component()
				p2 = list[unit].categories_to_string()
				self.warning("different category in same unit. (unit=[%s] comp/cate=[%s]/[%s] => [%s]/[%s])" % (unit, component, " ".jpin(category), p1, p2))
			return False
		return True

	def check_parent_unit(selt, list, unit, component, category) :
		d = os.path.dirname(unit)
		while d != "" :
			u = list.get(d)
			if u:
#				self.debug("check_parent_unit:; unit=[%s] component=[%s] d=[%s] u=[%s]" % (unit, component, d, u))
				if not u.equals_categories(category) :
					p1 = " ".join(category)
					self.warning("parent unit exists. (componment=[%s] unit=[%s] d=[%s] category=[%s])" % (component, unit, d, p1))
				else :
#					self.debug("exists_parent_unit: skip unit=[%s] component=[%s] d=[%s]" % (unit, component, d))
					pass
				return False
			d = os.path.dirname(d)
		return True

	def check_child_unit(self, list, unit, component, category) :
		rc = True
		for u in list.keys() :
			if u.startswith(unit + "/") :
				if not list[u]/equals_categories(category) :
					p1 = " ".join(category)
					self.warning("child unit exists. (component=[%s] unit=[%s] child=[%s] categor=[%s])" % (component, unit, u, p1))
					rc = False
				self.debug("check_child_unit: delete component=[%s] unit=[%s] child=[%s (%s)])" % (component, unit, u, list[u].get_component()))
				del list[u]
		return rc

	def find_unit(self, dir, component) :
		unit = self.STATIC_MAPPING.get(component)
		if unit :
			if not os.path.exists(unit) :
				print("Error: unit not exists. (unit=[%s] component=[%s])" % (unit, dir + unit))
				sys.exit(1)
			return unit

		paths = []
		find = subprocess.Popen(["/cygdrive/c/cygwin/bin/find.exe", dir, "-name", component, "-type", "f"], stdout = subprocess.PIPE)
		for path in iter(find.stdout.readline, b'') :
			path = path.rstrip("\n")
			self.debug("find_unit: push [%s]" % path)
			paths.append(path[len(dir) : ])

		if not paths :
			return ""

		srcdirs = []
		if len(paths) > 1 :
			for path in paths :
				if  ("/bin/" in path) \
				 or ("/doc/" in path) \
				 or (component.startswith("lib") and component.endswith(".so") and "/lib/" in path) :
					pass
				else :
					srcdirs.append(path)
			if len(srcdirs) > 1 :
				self.warning("duplicate paths. (component=[%s] paths=[%s])" % (component, " ".join(srcdirs)))
			elif len(srcdirs) < 1 :
				srcdirs = paths
		else :
			srcdirs = paths

		name = component
		if component.startswith("lib") and component.endswith(".so") :
			name = component[3 : len(component) - 3]

		srcdir = os.path.dirname(srcdirs[0])
		unit = ""
		if name == os.path.basename(srcdir) :
			unit = srcdir
		elif name == "Dss" + os.path.basename(srcdir) :
			unit = srcdir
		else :
			index = srcdir.find("src")
			if index > 0 :
				unit = srcdir[0 : index - 1]
			else :
				unit = srcdir

		if not os.path.exists(dir + unit) :
			print("Error: unit not exists. (unit=[%s] path=[%s])" % (unit, dir + unit))
			sys.exit(1)

#		self.debug("find_unit: dir=[%s] component=[%s] unit=[%s]" % (dir, component, unit))
		return unit

	def make_complist(self, path, topdir) :
		list = {}

		with open(path, "r") as ins :
#		with open(path, "r", encoding="cp932") as ins :
			for rec in ins :
				rec = rec.strip()
#				self.debug("make_complist: rec=[%s]" % rec)

				category = []
				component = ""
				result2 = self.PATTERN2.match(rec)
				if result2 :
					category = (result2.group(1), result2.group(2), result2.group(3), result2.group(4))
					component = result2.group(5)
				else :
					self.warning("complist file format. (rec=[%s])" % rec)
					continue

				if not component or component == "0" or component == "-" :
					continue
				if self.skip_component(component) :
					continue

				unit = self.find_unit(topdir, component)
				if unit == "" :
					self.warning("component not found. (component=[%s] category=[%s])" % (component, " ".join(category)))
					continue

				if not self.check_duplicate_unit(list, unit, component, category) :
					continue
				if not self.check_parent_unit(list, unit, component, category) :
					continue
				if not self.check_child_unit(list, unit, component, category) :
					continue

				self.debug("make_complist: add unit=[%s] component=[%s] category=[%s]" % (unit, component, " ".join(category)))
				list[unit] = category_component(category, component)
		return list

	def __init__(self, argv) :
		if (len(argv) != 3) :
			print "usage: python %s complist dir" % os.path.basename(argv[0])
			sys.exit(1)

		path = argv[2]
		if path[-1] != "/" :
			path += "/"

		list = self.make_complist(argv[1], path)
		print u"コンパイル単位\tシステムブロック\tサブシステム\tサービス群\tサービス\tコンポーネント"
		for unit in list :
			print "%s\t%s\t%s" % (unit, "\t".join(list[unit].get_category()), list[unit].get_component())

if __name__ == "__main__" :
#	sys.stdout = codecs.getwriter("utf-8")(sys.stdout)
	make_mapping(sys.argv)

File: src/mapping/make_mapping.pl
#!/usr/bin/perl
# -*- vi: set ts=4 sts=4 sw=4 : -*-

use strict;
#use warning;
use utf8;
use File::Basename;

#binmode STDOUT ":utf8";	# why?
#binmode STDERR ":utf8";	# why?

our %STATIC_MAPPING = (
	"..." => "...",
);

our @SKIP_COMPONENTS = (
	"...",
);

sub debug {
	my ($msg) = @_;
	print "$msg\n";
	return;
}

sub warning {
	my ($msg) = @_;
	print STDERR "Warning: $msg\n";
#	print "Warning: $msg\n";
	return;
}

sub categories_to_string {
	my ($c) = @_;
	return @$c[0]." ".@$c[1]." ".@$c[2]." ".@$c[3];
}

sub equals_categories {
	my ($c1, $c2) = @_;
	my $rc = @$c1[0] eq @$c2[0]
		&& @$c1[1] eq @$c2[1]	# "and" ではwarning...why?
		&& @$c1[2] eq @$c2[2]
		&& @$c1[3] eq @$c2[3]
		;
	return$rc;
}

sub get_component {
	my ($c) = @_;
	return @$c[4];
}

sub set_component {
	my ($category, $c) = @_;
	push @$category, $c;
}

sub skip_component {
	my ($c) = @_;

	if ($c =~ /\.(sh|adf|sts|ear|jar)$/) {
debug("skip_compoent: excluded component=[$c]");
		return 1;
	}
	for (@SKIP_COMPONENTS) {
		if ($_ eq $c) {
debug("skip_compoent: excluded component=[$c]");
			return 1;
		}
	}
	return 0;
}

sub check_duplicate_unit {
	my ($list2, $unit, $component, $cate) = @_;
	my %list = %$list2;
	my @category = @$cate;

	if ($list{$unit}) {
		if ($component eq get_component(\@{$list{$unit}})) {
			if (!equals_categories(\@category, \@{$list{$unit}})) {
				my $p1 = categories_to_string(\@category);
				my $p2 = categories_to_string(\@{$list{$unit}});
				warning("duplicate components. (component=[$component] category=[$p1] => [$p2])");
			}
			return 0;
		}
		if (!equals_categories(\@category, \@{$list{$unit}})) {
			my $p1 = get_component(\@{$list{$unit}});
			my $p2 = categories_to_string(\@{$list{$unit}});
			warning("different category in same unit. (unit=[$unit] comp/cate=[$component]/[@category] => [$p1]/[$p2])");
		}
		return 0;
	}
	return 1;
}

sub check_parent_unit {
	my ($list2, $unit, $component, $cate) = @_;
	my %list = %$list2;
	my @category = @$cate;

	my $d = $unit;
	while ($d ne ".") {
		$d = dirname $d;
		if ($list{$d}) {
			if (!equals_categories(\@category, \@{$list{$d}})) {
				my $p1 = categories_to_string(\@category);
				warning("parent unit exists. (component=[$component] unit=[$unit] parent=[$d] categopry=[$p1])");
			}
			else {
#debug "check_parent_unit: skip unit=[$unit] component=[$component] parent=[$d]";
			}
			return 0;
		}
	}
	return 1;
}

sub check_child_unit {
	my ($list2, $unit, $component, $cate) = @_;
	my %list = %$list2;
	my @category = @$cate;

	my $rc = 1;
	for (keys %list) {
		if (/^$unit\//) {
#debug "exists_child_unit: _=[$_] unit=[$unit] category=[".categories_to_string \@[$list{$_}]."]";
			if (!equals_categories(\@category, \@{list{$_}})) {
				my $p1 = categories_to_string(\@category);
				warning("child unit exists. (component=[$component] unit=[$unit] child=[$_] category=[$p1]");
				$rc = 0;
			}
debug("check_child_unit: delete componet=[$component] unit=[$unit] child=[$_ (".(get_component(\@{$list{$_}})).")]");
			delete $list2->{$_};
		}
	}
	return $rc;
}

sub find_unit {
	my ($dir, $component) = @_;

	my $unit = $STATIC_MAPPING{$component};
	if ($unit) {
		if (!-d $dir.$unit) {
			die "Error: unit not exists. (unit=[$unit] path=[".$dir.$unit."])\n";
		}
		return $unit;
	}

	my @paths = ();
	my $find = "find $dir -name $component -type f |";
	open FIND, $find or die "Error: $!";
	while (<FIND>) {
		chomp;
#debug "find_unit: push [$_]";
		$_ =~ s/^$dir//g;
		push @paths, $_;
	}
	close FIND;

	if (!@paths) {
		return "";
	}

	my @srcdirs = ();
	if ($#paths >= 0) {
		foreach (@paths) {
			if ((/\/(bin|doc)\//) or (/\/lib\// and $component =~ /^lib.*\.so$/)) {
			}
			else {
				push @srcdirs, $_;
			}
		}
		if ($#srcdirs > 0) {
			warning("duplicate files. (component=[$component] srcdirs=[@srcdirs])");
		}
		elsif ($#srcdirs < 0) {
			@srcdirs = @paths;
		}
	}

	my $name = $component;
	if ($component =~ /lib(\w*)\.so/) {
		$name = $1;
	}

	my $srcdir = dirname $srcdirs[0];
	my $unit = "";
	if ($name eq basename $srcdir) {
		$unit = $srcdir;
	}
	elsif ($name eq "Dss".basename $srcdir) {
		$unit = $srcdir;
	}
	else {
		my $ix = index $srcdir, "src";
		if ($ix > 0) {
			$unit = substr $srcdir, 0, $ix - 1;
		}
		else {
			$unit = $srcdir;
		}
	}

debug("find_unit: return unit=[$unit] component=[$component] srcdir=[$srcdir]");
	if (!-d $dir.$unit) {
		die "Error: unit not exists. (unit=[$unit])\n";
	}
	return $unit;
}

sub make_complist {
	my ($path, $topdir) = @_;
	my %list = ();

	open FILE, "<".$path or die "Error: $!";
	binmode FILE, ":crlf:encoding(cp932)";
	while (<FILE>) {
		chomp;
		my @category = ();
		my $component;
#debug "make_complist: rec=[$_]";
		if (/^([\S ]*)\t([\S ]*)\t([\S ]*)\t([\S ]*)\t([\S ]*)\t([\w\.-]*)\t*$/) {
			@category = ($1, $2, $3. $4);
			$component = $5;
		}
		else {
			warning("complist file format. (rec=[$_])");
			next;
		}
		next if (!$component or $component eq "0" or $component eq "-");
		next if (skip_component $component);

		my $unit = find_unit $topdir, $component;
		if (!$unit) {
			warning("component not found. (component=[$component] category=[@category])");
			next;
		}
		next if (!check_duplicate_unit(\%list, $unit, $component, \@category));
		next if (!check_parent_unit(\%list, $unit, $component, \@category));
		next if (!check_child_unit(\%list, $unit, $component, \@category));

debug("make_complist: add unit=[$unit] component=[$component] category=[@category]");
		set_component(\@category, $component);
		@{$list{$unit}} = @category;
	}
	close FILE;
	return %list;
}

sub main() {
	if ($#ARGV < 1) {
		print "usage: perl ".basename $0." complist dir\n";
		exit 1;
	}

	my $path = $ARGV[1];
	if (substr($path, -1) ne "/") {
		$path = $path."/";
	}
	my %list = make_complist $ARGV[1], $path;
	print "コンパイル単位\tシステムブロック\tサブシステム\tサービス群\tサービス\tコンポーネント\n";
	foreach (keys %list) {
		print $_."\t".join("\t", @{$list{$_}})."\n";
	}
}

&main;
